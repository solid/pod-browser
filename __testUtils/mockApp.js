/**
 * Copyright 2020 Inrupt Inc.
 *
 * Permission is hereby granted, free of charge, to any person obtaining a copy
 * of this software and associated documentation files (the "Software"), to deal in
 * the Software without restriction, including without limitation the rights to use,
 * copy, modify, merge, publish, distribute, sublicense, and/or sell copies of the
 * Software, and to permit persons to whom the Software is furnished to do so,
 * subject to the following conditions:
 *
 * The above copyright notice and this permission notice shall be included in
 * all copies or substantial portions of the Software.
 *
 * THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND, EXPRESS OR IMPLIED,
 * INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY, FITNESS FOR A
 * PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO EVENT SHALL THE AUTHORS OR COPYRIGHT
 * HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER LIABILITY, WHETHER IN AN ACTION
 * OF CONTRACT, TORT OR OTHERWISE, ARISING FROM, OUT OF OR IN CONNECTION WITH THE
 * SOFTWARE OR THE USE OR OTHER DEALINGS IN THE SOFTWARE.
 */

import {
  addStringNoLocale,
  addUrl,
  createSolidDataset,
  createThing,
  setStringNoLocale,
  setThing,
  setUrl,
} from "@inrupt/solid-client";
import { vcard, rdf, schema } from "rdf-namespaces";
import { chain } from "../src/solidClientHelpers/utils";
import { vcardExtras } from "../src/addressBook";

export const oidcPrefix = "http://www.w3.org/ns/solid/oidc#";
export const TOS_PREDICATE = `${oidcPrefix}tos_uri`;
export const POLICY_PREDICATE = `${oidcPrefix}policy_uri`;
export const CONTACTS_PREDICATE = `${oidcPrefix}contacts`;
export const LOGO_PREDICATE = `${oidcPrefix}logo_uri`;

export const APP_WEBID = "https://mockappurl.com";
export const APP_POLICY_URL = "https://mockappurl.com/privacy-policy";
export const APP_TOS_URL = "https://mockappurl.com/terms-of-service";

const contactThing = chain(
  createThing({ url: "https://mockappurl.com#contacts" }),
  (t) => addStringNoLocale(t, vcard.fn, "Example Contact"),
  (t) => addStringNoLocale(t, vcard.hasTelephone, "555-5555"),
  (t) => addStringNoLocale(t, vcard.hasEmail, "example@email.com"),
  (t) => addUrl(t, vcard.url, "https://examplewebsite.com")
);

const dataset = setThing(createSolidDataset(), contactThing);

export function mockApp() {
  return chain(
    createThing({ url: APP_WEBID }),
    (t) => addStringNoLocale(t, vcard.fn, "Mock App"),
    (t) => addUrl(t, vcardExtras("WebId"), APP_WEBID),
    (t) => addUrl(t, rdf.type, schema.SoftwareApplication), // just guessing what the type might be for apps
    (t) => addUrl(t, TOS_PREDICATE, APP_TOS_URL),
    (t) => addUrl(t, POLICY_PREDICATE, APP_POLICY_URL),
    (t) => addUrl(t, CONTACTS_PREDICATE, "https://mockappurl.com#contacts")
  );
}

export function mockAppDataset() {
  return setThing(dataset, mockApp());
}
