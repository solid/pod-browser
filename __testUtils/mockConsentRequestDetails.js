/**
 * Copyright 2020 Inrupt Inc.
 *
 * Permission is hereby granted, free of charge, to any person obtaining a copy
 * of this software and associated documentation files (the "Software"), to deal in
 * the Software without restriction, including without limitation the rights to use,
 * copy, modify, merge, publish, distribute, sublicense, and/or sell copies of the
 * Software, and to permit persons to whom the Software is furnished to do so,
 * subject to the following conditions:
 *
 * The above copyright notice and this permission notice shall be included in
 * all copies or substantial portions of the Software.
 *
 * THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND, EXPRESS OR IMPLIED,
 * INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY, FITNESS FOR A
 * PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO EVENT SHALL THE AUTHORS OR COPYRIGHT
 * HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER LIABILITY, WHETHER IN AN ACTION
 * OF CONTRACT, TORT OR OTHERWISE, ARISING FROM, OUT OF OR IN CONNECTION WITH THE
 * SOFTWARE OR THE USE OR OTHER DEALINGS IN THE SOFTWARE.
 */

async function getConsentRequestDetails() {
  return {
    "@context": [
      "https://www.w3.org/2018/credentials/v1",
      "https://w3id.org/security/suites/ed25519-2020/v1",
      "https://consent.pod.inrupt.com/credentials/v1",
    ],
    id: "https://consent.pod.inrupt.com/vc/53973727-f4d0-9e8dbdc041fd",
    type: ["VerifiableCredential", "SolidCredential", "SolidConsentRequest"],
    issuer: "https://consent.pod.inrupt.com",
    issuanceDate: "2021-05-26T16:40:03Z",
    expirationDate: "2021-06-23T16:40:03Z",
    credentialSubject: {
      id: "https://pod.inrupt.com/bob/profile/card#me",
      inbox: "https://pod.inrupt.com/bob/inbox/",
      hasConsent: {
        mode: ["Read", "Write"],
        hasStatus: "ConsentStatusRequested",
        forPersonalData: "https://pod.inrupt.com/alice/private/data",
        forPurpose: "https://example.com/SomeSpecificPurpose",
      },
    },
    proof: {
      created: "2021-05-26T16:40:03.009Z",
      proofPurpose: "assertionMethod",
      proofValue: "eqp8h_kL1DwJCpn65z-d1Arnysx6b11...jb8j0MxUCc1uDQ",
      type: "Ed25519Signature2020",
      verificationMethod: "https://consent.pod.inrupt.com/key/396f686b4ec",
    },
  };
}

export default getConsentRequestDetails;
