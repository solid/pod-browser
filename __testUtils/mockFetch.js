/**
 * Copyright 2020 Inrupt Inc.
 *
 * Permission is hereby granted, free of charge, to any person obtaining a copy
 * of this software and associated documentation files (the "Software"), to deal in
 * the Software without restriction, including without limitation the rights to use,
 * copy, modify, merge, publish, distribute, sublicense, and/or sell copies of the
 * Software, and to permit persons to whom the Software is furnished to do so,
 * subject to the following conditions:
 *
 * The above copyright notice and this permission notice shall be included in
 * all copies or substantial portions of the Software.
 *
 * THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND, EXPRESS OR IMPLIED,
 * INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY, FITNESS FOR A
 * PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO EVENT SHALL THE AUTHORS OR COPYRIGHT
 * HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER LIABILITY, WHETHER IN AN ACTION
 * OF CONTRACT, TORT OR OTHERWISE, ARISING FROM, OUT OF OR IN CONNECTION WITH THE
 * SOFTWARE OR THE USE OR OTHER DEALINGS IN THE SOFTWARE.
 */

export default function mockFetch(responses = {}) {
  return jest.fn().mockImplementation((url, init) => {
    const responseFn = responses[url.toString()];
    const responseObject = responseFn;
    const method = init?.method || "GET";
    let response;
    if (responseFn && typeof responseObject[method] !== "undefined") {
      response = responseObject[method](url, init);
    } else if (responseFn) {
      response = responseFn(url, init);
    } else {
      throw new Error(`URL (${url}) not mocked properly`);
    }
    Object.defineProperty(response, "url", { value: url });
    return Promise.resolve(response);
  });
}

export function mockSolidClientRequest(body, method = "GET", headers = {}) {
  return new Request(body, {
    headers: new Headers(headers),
    method,
  });
}

export function mockTurtleResponse(status, body = "") {
  return new Response(body, {
    status,
    headers: {
      "Content-Type": "text/turtle",
    },
  });
}
