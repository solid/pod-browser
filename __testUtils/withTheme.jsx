/**
 * Copyright 2020 Inrupt Inc.
 *
 * Permission is hereby granted, free of charge, to any person obtaining a copy
 * of this software and associated documentation files (the "Software"), to deal in
 * the Software without restriction, including without limitation the rights to use,
 * copy, modify, merge, publish, distribute, sublicense, and/or sell copies of the
 * Software, and to permit persons to whom the Software is furnished to do so,
 * subject to the following conditions:
 *
 * The above copyright notice and this permission notice shall be included in
 * all copies or substantial portions of the Software.
 *
 * THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND, EXPRESS OR IMPLIED,
 * INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY, FITNESS FOR A
 * PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO EVENT SHALL THE AUTHORS OR COPYRIGHT
 * HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER LIABILITY, WHETHER IN AN ACTION
 * OF CONTRACT, TORT OR OTHERWISE, ARISING FROM, OUT OF OR IN CONNECTION WITH THE
 * SOFTWARE OR THE USE OR OTHER DEALINGS IN THE SOFTWARE.
 */

import React from "react";
import { render } from "@testing-library/react";
import T from "prop-types";
import { StylesProvider, ThemeProvider } from "@inrupt/prism-react-components";
import { FeatureProvider } from "../src/contexts/featureFlagsContext";
import defaultTheme from "../src/theme";

export default function WithTheme(props) {
  const { theme, children } = props;
  return (
    <StylesProvider generateClassName={(rule) => `PodBrowser-${rule.key}`}>
      <ThemeProvider theme={theme}>{children}</ThemeProvider>
    </StylesProvider>
  );
}

WithTheme.propTypes = {
  // eslint-disable-next-line react/forbid-prop-types
  theme: T.object,
  children: T.node.isRequired,
};

WithTheme.defaultProps = {
  theme: defaultTheme,
};

export const renderWithTheme = (children, theme = defaultTheme) => {
  return render(
    <FeatureProvider>
      <WithTheme theme={theme}>{children}</WithTheme>
    </FeatureProvider>
  );
};
