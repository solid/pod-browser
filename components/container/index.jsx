/**
 * Copyright 2020 Inrupt Inc.
 *
 * Permission is hereby granted, free of charge, to any person obtaining a copy
 * of this software and associated documentation files (the "Software"), to deal in
 * the Software without restriction, including without limitation the rights to use,
 * copy, modify, merge, publish, distribute, sublicense, and/or sell copies of the
 * Software, and to permit persons to whom the Software is furnished to do so,
 * subject to the following conditions:
 *
 * The above copyright notice and this permission notice shall be included in
 * all copies or substantial portions of the Software.
 *
 * THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND, EXPRESS OR IMPLIED,
 * INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY, FITNESS FOR A
 * PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO EVENT SHALL THE AUTHORS OR COPYRIGHT
 * HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER LIABILITY, WHETHER IN AN ACTION
 * OF CONTRACT, TORT OR OTHERWISE, ARISING FROM, OUT OF OR IN CONNECTION WITH THE
 * SOFTWARE OR THE USE OR OTHER DEALINGS IN THE SOFTWARE.
 */

import React, { useEffect, useMemo, useState } from "react";
import T from "prop-types";
import { getSourceUrl, isContainer } from "@inrupt/solid-client";
import { useSession } from "@inrupt/solid-ui-react";
import { renderResourceType } from "../containerTableRow";
import { getResourceName } from "../../src/solidClientHelpers/resource";
import Spinner from "../spinner";
import PageHeader from "../containerPageHeader";
import ContainerDetails from "../containerDetails";
import { BookmarksContextProvider } from "../../src/contexts/bookmarksContext";
import AccessForbidden from "../accessForbidden";
import ResourceNotFound from "../resourceNotFound";
import useContainer from "../../src/hooks/useContainer";
import NotSupported from "../notSupported";
import { getContainerResourceUrlAll } from "../../src/models/container";
import { getContainerUrl } from "../../src/stringHelpers";
import ContainerSubHeader from "../containerSubHeader";
import useAuthenticatedProfile from "../../src/hooks/useAuthenticatedProfile";
import AuthProfileLoadError from "../authProfileLoadError";
import NoControlWarning from "../noControlWarning";
import useResourceInfo from "../../src/hooks/useResourceInfo";
import useAccessControl from "../../src/hooks/useAccessControl";
import PodRootLoadError from "../podRootLoadError";
import ContainerTable from "../containerTable";
import { isHTTPError } from "../../src/error";
import { locationIsConnectedToProfile } from "../../src/solidClientHelpers/profile";
import { isContainerIri } from "../../src/solidClientHelpers/utils";
import { downloadResource } from "../downloadLink";
import useAccessToResourceAndParentContainer from "../../src/hooks/useAccessToResourceAndParentContainer";
import DownloadResourceMessage from "../downloadResourceMessage";
import usePodRootUri from "../../src/hooks/usePodRootUri";
import NoPodFoundError from "../noPodFoundError";

function isNotAContainerResource(iri, container) {
  if (!iri) return true;
  return (
    container && isContainerIri(iri) && getSourceUrl(container.dataset) !== iri
  );
}

function renderContainerContent(isValidating, iri, data) {
  return isValidating ? (
    <Spinner />
  ) : (
    <ContainerTable
      containerPath={getContainerUrl(iri)}
      data={data}
      resourcePath={iri}
    />
  );
}

function maybeRenderWarning(locationIsInUsersPod, noControlError, podRootIri) {
  return (
    locationIsInUsersPod &&
    noControlError && <NoControlWarning podRootIri={podRootIri} />
  );
}

export default function Container({ iri }) {
  const [download, setDownload] = useState(false);
  const { sessionRequestInProgress, session } = useSession();
  const [resourceUrls, setResourceUrls] = useState(null);
  const authenticatedProfile = useAuthenticatedProfile();
  const podRootIri = usePodRootUri(iri);
  const [noControlError, setNoControlError] = useState(null);
  const { data: podRootResourceInfo, error: podRootError } =
    useResourceInfo(podRootIri);
  const {
    error: accessControlError,
    isValidating: validatingPodRootAccessControl,
  } = useAccessControl(podRootResourceInfo);

  const {
    data: container,
    error: containerError,
    mutate: update,
    isValidating,
  } = useContainer(iri);

  const { accessToParentContainer, accessToResource } =
    useAccessToResourceAndParentContainer(iri);

  useEffect(() => {
    if (accessToResource?.read && !accessToParentContainer?.read) {
      setDownload(true);
      downloadResource(iri, session.fetch);
    } else {
      setDownload(false);
    }
  }, [accessToParentContainer, accessToResource, iri, session]);

  useEffect(() => {
    if (isNotAContainerResource(iri, container)) return;
    const urls = container && getContainerResourceUrlAll(container);
    setResourceUrls(urls);
  }, [container, iri]);

  useEffect(() => {
    setNoControlError(accessControlError);
  }, [accessControlError]);

  const data = useMemo(() => {
    if (!resourceUrls) {
      return [];
    }

    return resourceUrls.map((rIri) => ({
      iri: rIri,
      name: getResourceName(rIri),
      filename: getResourceName(rIri)?.toLowerCase(),
      type: renderResourceType(rIri),
    }));
  }, [resourceUrls]);

  if (authenticatedProfile && podRootIri === null) return <NoPodFoundError />;

  const locationIsInUsersPod = locationIsConnectedToProfile(
    authenticatedProfile,
    iri
  );

  if (podRootError && locationIsInUsersPod) return <PodRootLoadError />;

  if ((!iri || isValidating || validatingPodRootAccessControl) && !podRootError)
    return <Spinner />;

  if (iri && download) {
    return <DownloadResourceMessage iri={iri} />;
  }

  if (containerError && isHTTPError(containerError.message, 401))
    return <AccessForbidden />;
  if (containerError && isHTTPError(containerError.message, 403))
    return <AccessForbidden />;
  if (containerError && isHTTPError(containerError.message, 404))
    return <ResourceNotFound />;
  if (containerError && !sessionRequestInProgress) return <NotSupported />;

  if (!authenticatedProfile) return <AuthProfileLoadError />;

  if (!resourceUrls || !container) return <Spinner />;

  const { dataset: containerDataset } = container;

  if (containerDataset && !isContainer(containerDataset))
    return <NotSupported />;

  return (
    <>
      <BookmarksContextProvider>
        <PageHeader />
        <ContainerDetails update={update}>
          <ContainerSubHeader update={update} resourceList={data} />
          {maybeRenderWarning(locationIsInUsersPod, noControlError, podRootIri)}
          {renderContainerContent(isValidating, iri, data)}
        </ContainerDetails>
      </BookmarksContextProvider>
    </>
  );
}

Container.propTypes = {
  iri: T.string,
};

Container.defaultProps = {
  iri: null,
};
