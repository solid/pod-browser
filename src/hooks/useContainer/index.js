/**
 * Copyright 2020 Inrupt Inc.
 *
 * Permission is hereby granted, free of charge, to any person obtaining a copy
 * of this software and associated documentation files (the "Software"), to deal in
 * the Software without restriction, including without limitation the rights to use,
 * copy, modify, merge, publish, distribute, sublicense, and/or sell copies of the
 * Software, and to permit persons to whom the Software is furnished to do so,
 * subject to the following conditions:
 *
 * The above copyright notice and this permission notice shall be included in
 * all copies or substantial portions of the Software.
 *
 * THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND, EXPRESS OR IMPLIED,
 * INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY, FITNESS FOR A
 * PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO EVENT SHALL THE AUTHORS OR COPYRIGHT
 * HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER LIABILITY, WHETHER IN AN ACTION
 * OF CONTRACT, TORT OR OTHERWISE, ARISING FROM, OUT OF OR IN CONNECTION WITH THE
 * SOFTWARE OR THE USE OR OTHER DEALINGS IN THE SOFTWARE.
 */

import useSWR from "swr";
import { useSession } from "@inrupt/solid-ui-react";
import { getContainerUrl } from "../../stringHelpers";
import { getContainer } from "../../models/container";

export default function useContainer(iri) {
  const session = useSession();
  const { sessionRequestInProgress } = session;
  const url = getContainerUrl(iri);

  async function fetchContainer(fetch) {
    return getContainer(url, { fetch });
  }

  return useSWR(
    ["container", url, sessionRequestInProgress],
    async () => {
      if (!url || sessionRequestInProgress) return null;
      return fetchContainer(session.fetch);
    },
    { revalidateOnFocus: false, refreshInterval: 0, errorRetryInterval: 2000 }
  );
}
