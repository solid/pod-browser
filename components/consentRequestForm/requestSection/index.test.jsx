/**
 * Copyright 2020 Inrupt Inc.
 *
 * Permission is hereby granted, free of charge, to any person obtaining a copy
 * of this software and associated documentation files (the "Software"), to deal in
 * the Software without restriction, including without limitation the rights to use,
 * copy, modify, merge, publish, distribute, sublicense, and/or sell copies of the
 * Software, and to permit persons to whom the Software is furnished to do so,
 * subject to the following conditions:
 *
 * The above copyright notice and this permission notice shall be included in
 * all copies or substantial portions of the Software.
 *
 * THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND, EXPRESS OR IMPLIED,
 * INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY, FITNESS FOR A
 * PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO EVENT SHALL THE AUTHORS OR COPYRIGHT
 * HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER LIABILITY, WHETHER IN AN ACTION
 * OF CONTRACT, TORT OR OTHERWISE, ARISING FROM, OUT OF OR IN CONNECTION WITH THE
 * SOFTWARE OR THE USE OR OTHER DEALINGS IN THE SOFTWARE.
 */

import React from "react";
import { waitFor } from "@testing-library/dom";
import userEvent from "@testing-library/user-event";
import { mockContainerFrom } from "@inrupt/solid-client";
import { renderWithTheme } from "../../../__testUtils/withTheme";
import RequestSection, {
  TESTCAFE_ID_REQUEST_SELECT_ALL_BUTTON,
  TESTCAFE_ID_REQUEST_EXPAND_SECTION_BUTTON,
  removeExistingValues,
} from "./index";
import useContainer from "../../../src/hooks/useContainer";
import * as containerFns from "../../../src/models/container";

jest.mock("../../../src/hooks/useContainer");
const mockedUseContainer = useContainer;

const sectionDetails = {
  mode: ["Read", "Write"],
  hasStatus: "ConsentStatusRequested",
  forPersonalData: ["https://pod.inrupt.com/alice/private/data/"],
  forPurpose: "https://example.com/SomeSpecificPurpose",
};

describe("Request Section", () => {
  beforeEach(() => {
    mockedUseContainer.mockReturnValue({
      data: {
        dataset: mockContainerFrom(
          "https://pod.inrupt.com/alice/private/data/"
        ),
      },
    });
    jest
      .spyOn(containerFns, "getContainerResourceUrlAll")
      .mockReturnValue([
        "https://pod.inrupt.com/alice/private/data/data-2",
        "https://pod.inrupt.com/alice/private/data/data-3",
      ]);
  });
  const setSelectedAccess = jest.fn();
  it("Renders initial context data", async () => {
    const { asFragment } = renderWithTheme(
      <RequestSection
        agentName="agent_name"
        sectionDetails={sectionDetails}
        setSelectedAccess={setSelectedAccess}
      />
    );

    expect(asFragment()).toMatchSnapshot();
  });

  it("selects all switches in a section and displays 'Deny all' in the toggle title", () => {
    const { getByTestId, getAllByRole, getByText } = renderWithTheme(
      <RequestSection
        agentName="agent_name"
        sectionDetails={sectionDetails}
        setSelectedAccess={setSelectedAccess}
      />
    );

    const selectAll = getByTestId(TESTCAFE_ID_REQUEST_SELECT_ALL_BUTTON);
    const switches = getAllByRole("checkbox");
    expect(getAllByRole("checkbox", { checked: false })).toHaveLength(
      switches.length
    );

    userEvent.click(selectAll);
    expect(getAllByRole("checkbox", { checked: true })).toHaveLength(
      switches.length
    );
    expect(selectAll).toHaveFocus();

    expect(getByText("Deny all")).toBeInTheDocument();
  });

  it("expands a sub-section and selects switch in it", async () => {
    const { getByTestId, getAllByRole } = renderWithTheme(
      <RequestSection
        agentName="agent_name"
        sectionDetails={sectionDetails}
        setSelectedAccess={setSelectedAccess}
      />
    );

    const expandSection = getByTestId(
      TESTCAFE_ID_REQUEST_EXPAND_SECTION_BUTTON
    );
    const switches = getAllByRole("checkbox", { checked: false });
    expect(getAllByRole("checkbox", { checked: false })).toHaveLength(
      switches.length
    );

    userEvent.click(expandSection);
    await waitFor(() => {
      expect(getAllByRole("checkbox")).toHaveLength(switches.length + 2);
    });

    const updatedSwitches = getAllByRole("checkbox", { checked: false });

    userEvent.click(updatedSwitches[0]);
    expect(getByTestId("consent-access-switch")).toBeChecked();
  });

  it("selects a single switch when clicked", async () => {
    const { getAllByRole, getByTestId } = renderWithTheme(
      <RequestSection
        agentName="agent_name"
        sectionDetails={sectionDetails}
        setSelectedAccess={setSelectedAccess}
      />
    );

    const switches = getAllByRole("checkbox");
    await waitFor(() => {
      expect(getAllByRole("checkbox", { checked: false })).toHaveLength(
        switches.length
      );
    });
    userEvent.click(switches[0]);
    await waitFor(() => {
      expect(getByTestId("consent-access-switch")).toBeChecked();
    });
  });

  it("switches all to unchecked when clicking deny all", () => {
    const { getByTestId, getAllByRole, getByText } = renderWithTheme(
      <RequestSection
        agentName="agent_name"
        sectionDetails={sectionDetails}
        setSelectedAccess={setSelectedAccess}
      />
    );

    const selectAll = getByTestId(TESTCAFE_ID_REQUEST_SELECT_ALL_BUTTON);
    const switches = getAllByRole("checkbox");
    expect(getAllByRole("checkbox", { checked: false })).toHaveLength(
      switches.length
    );

    userEvent.click(selectAll);
    expect(getAllByRole("checkbox", { checked: true })).toHaveLength(
      switches.length
    );
    expect(selectAll).toHaveFocus();

    const denyAllButton = getByText("Deny all");
    userEvent.click(denyAllButton);
    expect(getAllByRole("checkbox", { checked: false })).toHaveLength(
      switches.length
    );
  });
});

describe("removeExistingValues", () => {
  it("returns array of accesses after removing those checked to false in newer accesses and adding newly checked ones", () => {
    const previousAccess = [
      {
        index: 0,
        checked: true,
        accessModes: { read: true, write: false, append: false },
        resourceIri: "https://example.org",
      },
      {
        index: 1,
        checked: true,
        accessModes: { read: true, write: false, append: false },
        resourceIri: "https://example.org",
      },
    ];

    const newAccess = [
      {
        index: 0,
        checked: false,
        accessModes: { read: true, write: false, append: false },
        resourceIri: "https://example.org",
      },
      {
        index: 4,
        checked: true,
        accessModes: { read: true, write: false, append: false },
        resourceIri: "https://example.org",
      },
    ];

    const expectedResult = [
      {
        index: 1,
        checked: true,
        accessModes: { read: true, write: false, append: false },
        resourceIri: "https://example.org",
      },
      {
        index: 4,
        checked: true,
        accessModes: { read: true, write: false, append: false },
        resourceIri: "https://example.org",
      },
    ];
    expect(removeExistingValues(previousAccess, newAccess)).toEqual(
      expectedResult
    );
  });
});
