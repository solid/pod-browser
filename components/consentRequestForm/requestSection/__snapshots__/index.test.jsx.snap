// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Request Section Renders initial context data 1`] = `
<DocumentFragment>
  <fieldset
    class="PodBrowser-root PodBrowser-request-container__section"
  >
    <legend
      class="PodBrowser-request-container__header-text PodBrowser-request-container__header-text--small"
    >
      <span
        class="PodBrowser-icon-edit PodBrowser-icon-small"
      />
      <span
        class="PodBrowser-header__content"
      >
        agent_name wants to create, edit, or delete
      </span>
      <button
        class="PodBrowser-button PodBrowser-button--secondary PodBrowser-request-container__button--small"
        data-testid="request-select-all"
        type="button"
      >
        <span
          class="PodBrowser-button__text"
        >
          <span
            class="PodBrowser-root PodBrowser-caption"
          >
            Select all
          </span>
        </span>
      </button>
    </legend>
    <div
      aria-label="position"
      class="PodBrowser-root PodBrowser-request-container__section PodBrowser-request-container__section--box PodBrowser-row"
    >
      <label
        class="PodBrowser-root PodBrowser-box__content PodBrowser-labelPlacementStart"
      >
        <span
          class="PodBrowser-root"
        >
          <span
            aria-disabled="false"
            class="PodBrowser-root PodBrowser-root PodBrowser-root PodBrowser-switchBase PodBrowser-colorPrimary"
          >
            <span
              class="PodBrowser-label"
            >
              <input
                class="PodBrowser-input PodBrowser-input"
                type="checkbox"
                value="https://pod.inrupt.com/alice/private/data"
              />
              <span
                class="PodBrowser-thumb"
              />
            </span>
            <span
              class="PodBrowser-root"
            />
          </span>
          <span
            class="PodBrowser-track"
          />
        </span>
        <span
          class="PodBrowser-root PodBrowser-label PodBrowser-body1"
        >
          <p
            class="PodBrowser-root PodBrowser-body2"
          >
            <span
              class="PodBrowser-icon-file PodBrowser-icon-small PodBrowser-icon-small--dark"
            />
            <span
              class=""
              title="https://pod.inrupt.com/alice/private/data"
            >
              data
            </span>
          </p>
        </span>
      </label>
      <label
        class="PodBrowser-root PodBrowser-box__content PodBrowser-labelPlacementStart"
      >
        <span
          class="PodBrowser-root"
        >
          <span
            aria-disabled="false"
            class="PodBrowser-root PodBrowser-root PodBrowser-root PodBrowser-switchBase PodBrowser-colorPrimary"
          >
            <span
              class="PodBrowser-label"
            >
              <input
                class="PodBrowser-input PodBrowser-input"
                type="checkbox"
                value="https://pod.inrupt.com/alice/private/data-2/"
              />
              <span
                class="PodBrowser-thumb"
              />
            </span>
            <span
              class="PodBrowser-root"
            />
          </span>
          <span
            class="PodBrowser-track"
          />
        </span>
        <span
          class="PodBrowser-root PodBrowser-label PodBrowser-body1"
        >
          <p
            class="PodBrowser-root PodBrowser-body2"
          >
            <span
              class="PodBrowser-icon-folder PodBrowser-icon-small PodBrowser-icon-small--dark"
            />
            <span
              class=""
              title="https://pod.inrupt.com/alice/private/data-2/"
            >
              data-2
            </span>
            <button
              class="PodBrowser-dropdown-caret"
              data-testid="expand-section"
              type="button"
            >
              <span
                class="PodBrowser-icon-caret-down PodBrowser-icon-small--padded"
              />
            </button>
          </p>
        </span>
      </label>
      <label
        class="PodBrowser-root PodBrowser-box__content PodBrowser-labelPlacementStart"
      >
        <span
          class="PodBrowser-root"
        >
          <span
            aria-disabled="false"
            class="PodBrowser-root PodBrowser-root PodBrowser-root PodBrowser-switchBase PodBrowser-colorPrimary"
          >
            <span
              class="PodBrowser-label"
            >
              <input
                class="PodBrowser-input PodBrowser-input"
                type="checkbox"
                value="https://pod.inrupt.com/alice/private/data-3"
              />
              <span
                class="PodBrowser-thumb"
              />
            </span>
            <span
              class="PodBrowser-root"
            />
          </span>
          <span
            class="PodBrowser-track"
          />
        </span>
        <span
          class="PodBrowser-root PodBrowser-label PodBrowser-body1"
        >
          <p
            class="PodBrowser-root PodBrowser-body2"
          >
            <span
              class="PodBrowser-icon-file PodBrowser-icon-small PodBrowser-icon-small--dark"
            />
            <span
              class=""
              title="https://pod.inrupt.com/alice/private/data-3"
            >
              data-3
            </span>
          </p>
        </span>
      </label>
    </div>
  </fieldset>
</DocumentFragment>
`;
